File: backend/repositories/repositories/user_repository.py
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from sqlalchemy.exc import IntegrityError
from repositories.models.user import User, UserRole
from repositories.repositories.base_repository import BaseRepository
from pydantic import BaseModel
from typing import Optional
from uuid import UUID


# Pydantic schemas for CRUD operations
class UserCreate(BaseModel):
    """Schema for creating a user"""
    supabase_user_id: UUID
    username: str
    email: str
    first_name: str
    last_name: str
    role: UserRole

    class Config:
        from_attributes = True


class UserUpdate(BaseModel):
    """Schema for updating a user"""
    username: Optional[str] = None
    email: Optional[str] = None
    first_name: Optional[str] = None
    last_name: Optional[str] = None
    role: Optional[UserRole] = None

    class Config:
        from_attributes = True


class UserRepository(BaseRepository[User, UserCreate, UserUpdate]):
    """
    Repository pattern for User data access
    Inherits common CRUD operations from BaseRepository
    Adds user-specific query methods
    """

    def __init__(self, db: AsyncSession):
        super().__init__(User, db)

    async def create_user(
        self,
        supabase_user_id: UUID,
        username: str,
        email: str,
        first_name: str,
        last_name: str,
        role: UserRole
    ) -> Optional[User]:
        """
        Create a new user in the database

        Args:
            supabase_user_id: UUID from Supabase auth.users
            username: Unique username
            email: User's email address
            first_name: User's first name
            last_name: User's last name
            role: User role (student or instructor)

        Returns:
            Created User object or None if creation failed
        """
        user_create = UserCreate(
            supabase_user_id=supabase_user_id,
            username=username,
            email=email,
            first_name=first_name,
            last_name=last_name,
            role=role
        )
        return await self.create(user_create)

    async def get_user_by_id(self, user_id: int) -> Optional[User]:
        """
        Get user by internal database ID

        Args:
            user_id: Internal database user ID

        Returns:
            User object or None if not found
        """
        return await self.get(user_id)

    async def get_user_by_supabase_id(self, supabase_user_id: UUID) -> Optional[User]:
        """
        Get user by Supabase user ID

        Args:
            supabase_user_id: UUID from Supabase auth.users

        Returns:
            User object or None if not found
        """
        result = await self.db.execute(
            select(User).where(User.supabase_user_id == supabase_user_id)
        )
        return result.scalars().first()

    async def get_user_by_email(self, email: str) -> Optional[User]:
        """
        Get user by email address

        Args:
            email: User's email address

        Returns:
            User object or None if not found
        """
        result = await self.db.execute(
            select(User).where(User.email == email)
        )
        return result.scalars().first()

    async def get_user_by_username(self, username: str) -> Optional[User]:
        """
        Get user by username

        Args:
            username: User's username

        Returns:
            User object or None if not found
        """
        result = await self.db.execute(
            select(User).where(User.username == username)
        )
        return result.scalars().first()

    async def update_user(
        self,
        user_id: int,
        **kwargs
    ) -> Optional[User]:
        """
        Update user information

        Args:
            user_id: Internal database user ID
            **kwargs: Fields to update

        Returns:
            Updated User object or None if not found
        """
        # Filter out None values
        update_data = {k: v for k, v in kwargs.items() if v is not None}

        if not update_data:
            return await self.get_user_by_id(user_id)

        user_update = UserUpdate(**update_data)
        return await self.update(user_id, user_update)

    async def delete_user(self, user_id: int) -> bool:
        """
        Delete a user

        Args:
            user_id: Internal database user ID

        Returns:
            True if deleted, False if not found
        """
        return await self.delete(user_id)

    async def check_username_exists(self, username: str) -> bool:
        """
        Check if username already exists

        Args:
            username: Username to check

        Returns:
            True if exists, False otherwise
        """
        result = await self.db.execute(
            select(User.id).where(User.username == username)
        )
        return result.scalars().first() is not None

    async def check_email_exists(self, email: str) -> bool:
        """
        Check if email already exists

        Args:
            email: Email to check

        Returns:
            True if exists, False otherwise
        """
        result = await self.db.execute(
            select(User.id).where(User.email == email)
        )
        return result.scalars().first() is not None

File: backend-ts/src/repositories/user.repository.ts
import { eq, or, ilike, and, desc, count } from "drizzle-orm"
import { users, type User, type NewUser } from "../models/index.js"
import { BaseRepository } from "./base.repository.js"
import { injectable } from "tsyringe"

/** Valid user roles - single source of truth for both type and runtime validation */
export const USER_ROLES = ["student", "teacher", "admin"] as const

/** User role type - derived from the USER_ROLES array */
export type UserRole = (typeof USER_ROLES)[number]

/** Data required to create a new user */
export interface CreateUserData {
  supabaseUserId: string
  email: string
  firstName: string
  lastName: string
  role: UserRole
}

/** Data for updating an existing user */
export interface UpdateUserData {
  email?: string
  firstName?: string
  lastName?: string
  role?: UserRole
  avatarUrl?: string
}

/** Filter options for user queries */
export interface UserFilterOptions {
  page: number
  limit: number
  search?: string
  role?: UserRole
  status?: "active" | "inactive"
}

/** Paginated result structure */
export interface PaginatedResult<T> {
  data: T[]
  total: number
  page: number
  limit: number
  totalPages: number
}

/**
 * Repository for user-related database operations.
 * Provides methods for user CRUD and lookups.
 */
@injectable()
export class UserRepository extends BaseRepository<
  typeof users,
  User,
  NewUser
> {
  constructor() {
    super(users)
  }

  /** Create a new user */
  async createUser(data: CreateUserData): Promise<User> {
    const results = await this.db
      .insert(users)
      .values({
        supabaseUserId: data.supabaseUserId,
        email: data.email,
        firstName: data.firstName,
        lastName: data.lastName,
        role: data.role,
      })
      .returning()

    return results[0]
  }

  /** Get user by internal database ID */
  async getUserById(userId: number): Promise<User | undefined> {
    return await this.findById(userId)
  }

  /** Get user by Supabase user ID */
  async getUserBySupabaseId(supabaseUserId: string): Promise<User | undefined> {
    const results = await this.db
      .select()
      .from(users)
      .where(eq(users.supabaseUserId, supabaseUserId))
      .limit(1)

    return results[0]
  }

  /** Get user by email address */
  async getUserByEmail(email: string): Promise<User | undefined> {
    const results = await this.db
      .select()
      .from(users)
      .where(eq(users.email, email))
      .limit(1)

    return results[0]
  }

  /** Update user information */
  async updateUser(
    userId: number,
    data: UpdateUserData,
  ): Promise<User | undefined> {
    // Filter out undefined values
    const updateData = Object.fromEntries(
      Object.entries(data).filter(([_, v]) => v !== undefined),
    )

    if (Object.keys(updateData).length === 0) {
      return await this.getUserById(userId)
    }

    return await this.update(userId, updateData)
  }

  /** Delete a user */
  async deleteUser(userId: number): Promise<boolean> {
    return await this.delete(userId)
  }

  /** Check if email already exists */
  async checkEmailExists(email: string): Promise<boolean> {
    const results = await this.db
      .select({ id: users.id })
      .from(users)
      .where(eq(users.email, email))
      .limit(1)

    return results.length > 0
  }

  /**
   * Get all users with pagination and filters.
   * Moved from AdminService to follow DIP.
   */
  async getAllUsersFiltered(
    options: UserFilterOptions,
  ): Promise<PaginatedResult<User>> {
    const { page, limit, search, role, status } = options
    const offset = (page - 1) * limit

    // Build where conditions
    const conditions: ReturnType<typeof eq>[] = []

    if (search) {
      conditions.push(
        or(
          ilike(users.email, `%${search}%`),
          ilike(users.firstName, `%${search}%`),
          ilike(users.lastName, `%${search}%`),
        )!,
      )
    }

    if (role) {
      conditions.push(eq(users.role, role))
    }

    if (status === "active") {
      conditions.push(eq(users.isActive, true))
    } else if (status === "inactive") {
      conditions.push(eq(users.isActive, false))
    }

    const whereClause = conditions.length > 0 ? and(...conditions) : undefined

    // Get total count
    const countResult = await this.db
      .select({ count: count() })
      .from(users)
      .where(whereClause)

    const total = Number(countResult[0]?.count ?? 0)

    // Get paginated data
    const data = await this.db
      .select()
      .from(users)
      .where(whereClause)
      .orderBy(desc(users.createdAt))
      .limit(limit)
      .offset(offset)

    return {
      data,
      total,
      page,
      limit,
      totalPages: Math.ceil(total / limit),
    }
  }

  /**
   * Toggle user active status.
   * Moved from AdminService to follow DIP.
   */
  async toggleActiveStatus(userId: number): Promise<User | undefined> {
    const user = await this.findById(userId)
    if (!user) return undefined

    const newStatus = !user.isActive
    const results = await this.db
      .update(users)
      .set({ isActive: newStatus })
      .where(eq(users.id, userId))
      .returning()

    return results[0]
  }

  /** Get all users by role */
  async getUsersByRole(role: UserRole): Promise<User[]> {
    return await this.db
      .select()
      .from(users)
      .where(eq(users.role, role))
      .orderBy(desc(users.createdAt))
  }

  /**
   * Get user counts grouped by role.
   * Used for admin analytics dashboard.
   */
  async getCountsByRole(): Promise<Record<string, number>> {
    const results = await this.db
      .select({
        role: users.role,
        count: count(),
      })
      .from(users)
      .groupBy(users.role)

    const roleMap: Record<string, number> = {}
    results.forEach((row) => {
      roleMap[row.role] = Number(row.count)
    })
    return roleMap
  }

  /**
   * Get most recent users.
   * Used for admin activity feed.
   */
  async getRecentUsers(limit: number = 10): Promise<User[]> {
    return await this.db
      .select()
      .from(users)
      .orderBy(desc(users.createdAt))
      .limit(limit)
  }
}
